### PR TITLE
feat!: assert suggestion messages are unique in rule testers

### LINKED ISSUE #16908: Change Request: [RuleTester] error if multiple suggestion fixers have the same message
### ESLint version

latest

### What problem do you want to solve?

Currently it's possible to accidentally create a rule that emits two suggestion fixers with the same message. This is a problem because IDEs will show a list of applicable fixers using the suggestion message; so if two have the exact same message then it means there will be two items in the list with exactly the same label - which is obviously confusing to the user.

### What do you think is the correct solution?

I propose that `RuleTester` be updated so that it enforces that for any given error, the hydrated `desc` of all suggestion fixers must be a unique string, enforcing that there are no duplicates.

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

```js
// ❌ BAD
context.report({
  node,
  messageId: 'topLevel',
  suggest: [
    {
      fix: fixerOne,
      messageId: 'suggestionOne',
      data: { placeholder1: 'value1' },
    },
    {
      fix: fixerTwo,
      messageId: 'suggestionOne', // ❌ oops
      data: { placeholder1: 'value1' },
    },
  ],
});
```
```js
// ✅ GOOD
context.report({
  node,
  messageId: 'topLevel',
  suggest: [
    {
      fix: fixerOne,
      messageId: 'suggestionOne',
      data: { placeholder1: 'value1' },
    },
    {
      fix: fixerTwo,
      messageId: 'suggestionTwo', // ✅
      data: { placeholder1: 'value1' },
    },
  ],
});
```

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Added logic to the two rule tester classes to assert that no suggestion contains the same hydrated description as a previous suggestion in the same error. 

#### Is there anything you'd like reviewers to focus on?

Was the consensus for asserting on the rule's _actual_ reported suggestions or _expected_? I went with _actual_ to start because that feels more comprehensive. Relevant: #15104

Fixes #16908.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
#16908 is ready for review. Marked as draft for now as it's a breaking change.

**Review by nzakas — APPROVED** [review]
LGTM. And yes, the consensus was to track the actual messages produced from a rule.

We'll need to hold off on this until we start merging PRs for v9.

File: tests/lib/rule-tester/rule-tester.js
Comment (mdjermanovic) [inline tests/lib/rule-tester/rule-tester.js:2264]: This change doesn't seem necessary?

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
Just https://github.com/eslint/eslint/pull/17532#discussion_r1382559056, then LGTM.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
